File: Java/math/CatalanNumber.java
import java.util.*;

public class CatalanNumber {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        System.out.print("Enter number: ");
        int num = sc.nextInt();
        System.out.println();
        int[] arr = new int[num + 2];
        arr[0] = 1;
        arr[1] = 1;
        for (int i = 1; i <= num; i++) {
            for (int j = i; j > 1; j--)
                arr[j] = arr[j] + arr[j - 1];
            arr[i + 1] = arr[i];
            for (int j = i + 1; j > 1; j--)
                arr[j] = arr[j] + arr[j - 1];
            System.out.printf("%d ", arr[i + 1] - arr[i]);
        }
    }
}

/*
Sample input and output
Enter number: 7

1 2 5 14 42 132 429 

time complexity: O(n^2)
space complexity: O(n)
*/

File: Julia/math/catalan_number.jl
# Julian program to Find the n'th Catalan number
# To know more about Catalan Numbers Visit Here : https://en.wikipedia.org/wiki/Catalan_number


function find_catalan(n)
    if (n == 0 || n == 1)
        return 1
    end

    # DP Table 
    cat = [0 for i in 1:(n + 1)]

    # Initialize the first to values
    cat[1] = 1
    cat[2] = 1
    for i in 3:(n + 1)
        for j in 1:(i - 1)
            cat[i] = cat[i] +  (cat[j] * cat[i - j])
        end
    end
    return cat[n + 1]
end

print("Enter the number: ")
num = readline()
num = parse(Int, num)
catalan_number = find_catalan(abs(num))
println("The $num'th catalan number is $catalan_number.")

"""
Time Complexity: O(num^2), where 'num' is the given number
Space Complexity: O(num)

SAMPLE INPUT AND OUTPUT

SAMPLE 1
Enter the number: 12
The 12'th catalan number is 208012.

SAMPLE 2
Enter the number: 2
The 2'th catalan number is 2.

"""
